src/types.rs

```diff
@@ -226,22 +229,40 @@ impl PublicId {
         public_key : id.get_public_key(),
         public_sign_key : id.get_public_sign_key(),
         validation_token : id.get_validation_token(),
-        name : id.get_name()
+        name : id.get_name().clone(),
+        relocated : id.relocated,
       }
     }
 
+    pub fn name(&self) -> NameType {
+      self.name.clone()
+    }
+
     pub fn serialised_contents(&self)->Vec<u8> {
         let mut e = cbor::Encoder::from_memory();
         e.encode(&[&self]).unwrap();
         e.into_bytes()
     }
+
+    pub fn is_relocated(&self) -> bool {
+        self.relocated
+    }
+
+    pub fn assign_relocated_name(&mut self, relocated_name: NameType) {
+        debug_assert!(!self.relocated, "already assigned relocated name !");
+        debug_assert!(self.name != relocated_name, "using original name as relocated name !");
```

Comment:
this is a hard condition, where name should be recalculated, and if the stored name is already different from the calculated name, refuse re-assigning it.  The function is best to return the resulting name